src/circuit/gadget/ecc.rs

```diff
@@ -12,129 +12,227 @@ use halo2::{
 pub trait FixedPoints<C: CurveAffine>: Clone + fmt::Debug {}
 
 /// The set of circuit instructions required to use the ECC gadgets.
-pub trait EccInstructions<C: CurveAffine>: Chip<Field = C::Base> {
-    /// Variable representing an element of the elliptic curve's scalar field.
-    type Scalar: Clone + fmt::Debug;
+pub trait EccInstructions<C: CurveAffine>: Chip<C::Base> {
+    /// Variable representing a full-width element of the elliptic curve's scalar field, to be used for fixed-base scalar mul.
+    type ScalarFixed: Clone + fmt::Debug;
+    /// Variable representing a signed short element of the elliptic curve's scalar field, to be used for fixed-base scalar mul.
+    type ScalarFixedShort: Clone + fmt::Debug;
     /// Variable representing an elliptic curve point.
     type Point: Clone + fmt::Debug;
+    /// Variable representing the x-coordinate of an elliptic curve point.
+    type X: Clone + fmt::Debug;
     /// Variable representing the set of fixed bases in the circuit.
     type FixedPoints: FixedPoints<C>;
     /// Variable representing a fixed elliptic curve point (constant in the circuit).
     type FixedPoint: Clone + fmt::Debug;
 
-    /// Witnesses the given scalar as a private input to the circuit.
-    fn witness_scalar(
-        layouter: &mut impl Layouter<Self>,
+    /// Witnesses the given full-width scalar as a private input to the circuit for fixed-based scalar mul.
+    fn witness_scalar_fixed(
+        &self,
+        layouter: &mut impl Layouter<C::Base>,
+        value: Option<C::Scalar>,
+    ) -> Result<Self::ScalarFixed, Error>;
+
+    /// Witnesses the given signed short scalar as a private input to the circuit for fixed-based scalar mul.
+    fn witness_scalar_fixed_short(
+        &self,
+        layouter: &mut impl Layouter<C::Base>,
         value: Option<C::Scalar>,
-    ) -> Result<Self::Scalar, Error>;
+    ) -> Result<Self::ScalarFixedShort, Error>;
 
     /// Witnesses the given point as a private input to the circuit.
     fn witness_point(
-        layouter: &mut impl Layouter<Self>,
+        &self,
+        layouter: &mut impl Layouter<C::Base>,
         value: Option<C>,
     ) -> Result<Self::Point, Error>;
 
+    /// Extracts the x-coordinate of a point.
+    fn extract_p(point: &Self::Point) -> &Self::X;
+
     /// Gets a fixed point into the circuit.
-    fn get_fixed(
-        layouter: &mut impl Layouter<Self>,
```

Comment:
I presume the removal of the layouter from this API means it turned out to not be necessary.

Reply:
`get_fixed()` does not need to take a layouter because it doesn't assign any new cells. Instead, it returns references to pre-loaded cells, which can be copied by the caller into whichever part of the circuit they need.